File: src/main/source/org/glavo/viewer/util/PathUtils.java
package org.glavo.viewer.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.*;
import java.nio.file.FileSystemNotFoundException;
import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.Objects;

public final class PathUtils {
    private PathUtils() {
    }

    @NotNull
    public static String fileNameOf(@NotNull URI uri) {
        String[] s = uri.toString().split("/");
        return s[s.length - 1];
    }

    @NotNull
    public static String fileNameOf(@NotNull URL url) {
        String[] s = url.toString().split("/");
        return s[s.length - 1];
    }

    @Nullable
    public static String classNameOf(@NotNull URI uri) {
        String fn = fileNameOf(uri);
        if (fn.toLowerCase().equals(".class")) {
            return fn.substring(0, fn.length() - 6);
        } else {
            return null;
        }
    }

    @Nullable
    public static String classNameOf(@NotNull URL url) {
        String fn = fileNameOf(url);
        if (fn.toLowerCase().equals(".class")) {
            return fn.substring(0, fn.length() - 6);
        } else {
            return null;
        }
    }

    @NotNull
    public static URI toUri(@NotNull Path path) {
        try {
            return new URI(URLDecoder.decode(path.toUri().toString(), "UTF-8"));
        } catch (URISyntaxException | UnsupportedEncodingException e) {
            assert false;
            throw new RuntimeException(e);
        }
    }

    @NotNull
    public static URL toUrl(@NotNull Path path) {
        try {
            return new URL(URLDecoder.decode(path.toUri().toString(), "UTF-8"));
        } catch (MalformedURLException | UnsupportedEncodingException e) {
            assert false;
            throw new RuntimeException(e);
        }
    }

    @Nullable
    public static Path toPath(@NotNull URI uri) {
        Objects.requireNonNull(uri);
        try {
            return Paths.get(uri);
        } catch (FileSystemNotFoundException e) {
            try {
                FileSystems.newFileSystem(uri, Collections.<String, Object>emptyMap());
                return Paths.get(uri);
            } catch (IOException e1) {
                return null;
            }
        }
    }

    @Nullable
    public static Path toPath(@NotNull URL url) {
        try {
            return toPath(url.toURI());
        } catch (URISyntaxException e) {
            return null;
        }
    }
}

File: src/main/java/org/glavo/viewer/util/PathUtils.kt
@file:JvmName("UrlUtils")

package org.glavo.viewer.util

import java.net.URI
import java.net.URL
import java.net.URLDecoder
import java.nio.file.*

fun URL.readData(): ByteArray {
    return this.readBytes()
}

fun URI.readData(): ByteArray {
    return this.toURL().readBytes()
}

val URL.fileName: String
    get() = this.toString().split("/").last()

val URI.fileName: String
    get() = this.toString().split("/").last()

val URL.className: String?
    get() {
        val name = this.fileName
        return if (fileName.endsWith(".class", ignoreCase = true)) {
            name.substring(0, name.length - 6)
        } else {
            null
        }
    }

@JvmName("pathToURI")
fun Path.toURI(): URI {
    return URI(URLDecoder.decode(this.toUri().toString(), "UTF-8"))
}

@JvmName("pathToURL")
fun Path.toURL(): URL {
    return URL(URLDecoder.decode(this.toUri().toString(), "UTF-8"))
}

@JvmName("uriToPath")
fun URI.toPath(): Path? {
    return when (this.scheme) {
        "jar" -> {
            try {
                FileSystems.getFileSystem(this)
                Paths.get(this)
            } catch (e: FileSystemNotFoundException) {
                try {
                    FileSystems.newFileSystem(this, mapOf<String, Any>())
                    Paths.get(this)
                } catch (_: FileSystemNotFoundException) {
                    null
                } catch (_: FileSystemAlreadyExistsException) {
                    null
                }
            }
        }
        else -> try {
            Paths.get(this)
        } catch (_: FileSystemNotFoundException) {
            null
        }
    }
}

@JvmName("urlToPath")
fun URL.toPath(): Path? = this.toURI().toPath()